### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by bigdinotech**
-includes python script for generating .bin file that contains product
data section of OTP
-includes script for flashing .bin file int OTP section of flash using
JTAG

### CODE REVIEW

**Comment by bigdinotech** [comment]
@SidLeung @eriknyquist please review
@noelpaz @russmcinnis please test

**Comment by bigdinotech** [comment]
@eriknyquist please review
README will go on top of this commit
https://github.com/bigdinotech/CODK-Tools/tree/instructions_readme

File: otp_provisioning/generate_factory_bin_product.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+VERSIONHI = 2
+VERSIONLO = 0
+KEYSTART = 0xA5A5A5A5
+KEYEND = 0x5A5A5A5A
+
+def main():
+# Parse input arguments
+    parser = argparse.ArgumentParser(description="Generate a binary file containing"
+        "various provisioning product info such as product serial number.")
+    parser.add_argument('--product_sn', help="the product serial number, up to 16 ASCII characters", required=True)
```
Comment (eriknyquist) [inline otp_provisioning/generate_factory_bin_product.py:54]: "product" is superflous..... can you remove it from all the parameter names? i.e. --product-sn becomes --sn, or --serial

File: otp_provisioning/generate_factory_bin_product.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    parser.add_argument('--product_hw_ver', help="the product hardware version, up to 4 bytes in hexadecimal", required=True)
+    parser.add_argument('--product_board_name', help="the product board name", required=True)
```
Comment (eriknyquist) [inline otp_provisioning/generate_factory_bin_product.py:56]: Let's use the expected conventions here, to be curteous to the user.

1) add short versions of each argument, i.e. "-v", "--hw-ver"
2) use a dash as a word separator (not underscore)

File: otp_provisioning/generate_factory_bin_product.py (hunk 1 above)
Comment (bigdinotech) [reply otp_provisioning/generate_factory_bin_product.py:54]: I think we should keep it. That is from the original file. And really just added the parameters for VID/PID

File: otp_provisioning/generate_factory_bin_product.py (hunk 2 above)
Comment (bigdinotech) [reply otp_provisioning/generate_factory_bin_product.py:56]: same as above.

File: otp_provisioning/generate_factory_bin_product.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    #put the ending block pattern keys and version numbers
+    product_data.patternKeyStart = KEYSTART
+    product_data.blockVersionHi = VERSIONHI
+    product_data.blockVersionLo = VERSIONLO
+    product_data.patternKeyEnd = KEYEND
+
+# Save the .bin file
+    arr = bytearray(product_data)
+    assert len(arr) == 512, "ERROR: product data struct size error"
+    out_file = open("factory_data_product.bin", "wb")
```
Comment (eriknyquist) [inline otp_provisioning/generate_factory_bin_product.py:132]: Give it a name and stick it up top with MAGIC and VERSION (the filename string, I mean)....

File: otp_provisioning/generate_factory_bin_product.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+    out_file.close()
+
+
+# Creates the OpenOCD flash script
+    flash_script = """### Generated script, please do not edit
+### Flash product provisining data in OTP
+load_image factory_data_product.bin %s
+verify_image factory_data_product.bin %s
+""" % (base_addr, base_addr)
+    f = open('factory_data_product.cmd', 'w')
```
Comment (eriknyquist) [inline otp_provisioning/generate_factory_bin_product.py:143]: Same as the other filename...

File: otp_provisioning/generate_factory_bin_product.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+import argparse
+import ctypes
+import binascii
+
+import factory_data
+
+base_addr = hex(0xffffe000 + 0x200)
+
+GENUINO = "Genuino 101"
+ARDUINO = "Arduino 101"
```
Comment (eriknyquist) [inline otp_provisioning/generate_factory_bin_product.py:42]: GENUINO and ARDUINO don't seem to be used anywhere....

File: otp_provisioning/generate_factory_bin_product.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+VENDOR = "Intel"
```
Comment (eriknyquist) [inline otp_provisioning/generate_factory_bin_product.py:43]: VENDOR is also not used anywhere

File: otp_provisioning/generate_factory_bin_product.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    product_data = factory_data.CustomerData()
+
+    # Add the product serial number
+    ############
+    assert len(args.product_sn) <= 16
+
+    product_sn_padded = args.product_sn.ljust(16, '\0')
+    product_data.product_sn = (ctypes.c_ubyte * 16).from_buffer_copy(bytearray(product_sn_padded))
+
+    print "Product S/N     : " + product_sn_padded
```
Comment (eriknyquist) [inline otp_provisioning/generate_factory_bin_product.py:73]: These seem like debug prints.... all the printing of the final padded values. Did you mean to remove them?

File: otp_provisioning/generate_factory_bin_product.py [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+    parser.add_argument('--product_PID', help="the product PID", required=True)
+    args = parser.parse_args()
+
+
+# Create instance of customer data struct
+    product_data = factory_data.CustomerData()
+
+    # Add the product serial number
+    ############
+    assert len(args.product_sn) <= 16
```
Comment (eriknyquist) [inline otp_provisioning/generate_factory_bin_product.py:68]: Give them all names and put them up top. 4, 8, 16, 32, all of them

**Review by eriknyquist — CHANGES_REQUESTED** [review]
needs cleanup before it can be given to the general public...

File: otp_provisioning/generate_factory_bin_product.py (hunk 1 above)
Comment (eriknyquist) [reply otp_provisioning/generate_factory_bin_product.py:54]: Yes, but up until now, this has been a private factory script. Now, we are making it public and giving it out to everyone; we can do better than this

**Comment by russmcinnis** [comment]
I reviewed the instructions and they look clear enough for me. Nothing but the readme in there though.

**Comment by eriknyquist** [comment]
closing. This is replaced by #2

**Pull request closed without merging**